### PR TITLE
Add probes for critical, warning and change events in specified window with delay

### PR DESCRIPTION
**Pull request by gtfisher**
Signed-off-by: Grant <grant@Grants-MacBook-Pro.local>

Closes Issue 10

### CODE REVIEW

File: chaosinstana/api.py [hunk 1]
```diff
@@ -43,6 +43,6 @@ def get_all_events(from_time: str, to_time: str,
                 secrets.get("instana_api_token"))
         }
     )
-    logger.debug("r.json(): {}".format(r.json()))
+    # logger.debug("r.json(): {}".format(r.json()))
```
Comment (russmiles) [inline chaosinstana/api.py]: If this is not needed, it should be removed.

File: chaosinstana/probes.py [hunk 2]
```diff
@@ -1,13 +1,66 @@
 import dateparser
+# import datetime
```
Comment (russmiles) [inline chaosinstana/probes.py]: Same here, remove not comment out.

File: chaosinstana/probes.py [hunk 3]
```diff
@@ -52,3 +108,27 @@ def convert_time(source_time: str) -> int:
             logger.error("Invalid date string provided")
             raise ActivityFailed("Invalid date string provided")
         return int(start_dt.timestamp() * 1000)
+
+
+def has_severity(event_data, severity_level):
+    """
+    Loops thorugh event data (expecting json array) and
```
Comment (russmiles) [inline chaosinstana/probes.py]: Small typo in "through"

**Review by russmiles — CHANGES_REQUESTED** [review]
Some very small changes to consider.

File: chaosinstana/probes.py [hunk 4]
```diff
@@ -1,13 +1,66 @@
 import dateparser
+# import datetime
 import time
 
+
```
Comment (Lawouach) [inline chaosinstana/probes.py]: no need for these double blank lines between imports :)

File: chaosinstana/probes.py [hunk 5]
```diff
@@ -1,13 +1,66 @@
...
+WARNING_LEVEL = 5
+CHANGE_LEVEL = -1
+
+
+def has_critical_events_in_window(from_time: str = None, to_time: str = None,
+                                  delay: int = 0,
+                                  configuration: Configuration = None,
+                                  secrets: Secrets = None) -> str:
+    """
+    returns a boolean indicating if there have been any critical events in a
```
Comment (Lawouach) [inline chaosinstana/probes.py]: It's good etiquette to start with an upper case docstring sentences :)

File: chaosinstana/probes.py [hunk 6]
```diff
@@ -18,15 +71,18 @@ def get_all_events_in_window(from_time: str = None, to_time: str = None,
...
-    logger.debug("instana probe delay is: {}".format(delay))
+    logger.info("Instana probe delay is: {} seconds.".format(delay))
     time.sleep(delay)
-    from_time = convert_time(from_time)
-    to_time = convert_time(to_time)
-    logger.debug("from_time {}".format(from_time))
-    logger.debug("to_time {}".format(to_time))
+    from_milli = convert_time(from_time)
+    to_milli = convert_time(to_time)
+    # fr_date = datetime.datetime.fromtimestamp(int(from_milli)/1000.0)
```
Comment (Lawouach) [inline chaosinstana/probes.py]: remove if not needed :)

File: chaosinstana/probes.py [hunk 7]
```diff
@@ -52,3 +108,27 @@ def convert_time(source_time: str) -> int:
             logger.error("Invalid date string provided")
             raise ActivityFailed("Invalid date string provided")
         return int(start_dt.timestamp() * 1000)
+
+
+def has_severity(event_data, severity_level):
```
Comment (Lawouach) [inline chaosinstana/probes.py]: it would be good if the two parameters had a type annotation

Comment (Lawouach) [reply chaosinstana/probes.py]: and the function a return type annotation, or

```
from typing import NoReturn

def .... -> NoReturn:
...

````

when a function does not return anything explicitely

File: chaosinstana/probes.py [hunk 8]
```diff
@@ -52,3 +108,27 @@ def convert_time(source_time: str) -> int:
...
+            if severity == severity_level:
+                logger.debug("has_severity for severity_level: {} is True"
+                             .format(severity_level))
+                return True
+
+        logger.debug("has_severity for severity_level: {} is False"
+                     .format(severity_level))
+        return False
+    except ValueError:
+        logger.error("has_severity error parsing event_data")
```
Comment (Lawouach) [inline chaosinstana/probes.py]: shouldn't this be raising an ActivityFailed? I'm not sure.

Comment (Lawouach) [reply chaosinstana/probes.py]: but the log should be a debug one considering the user cannot action it

File: tests/test_probes.py [hunk 9]
```diff
@@ -5,6 +5,8 @@
 
 from chaosinstana.probes import get_all_events_in_window
 from chaosinstana.probes import convert_time
+from chaosinstana.probes import has_severity
```
Comment (Lawouach) [inline tests/test_probes.py]: we tend to regroup all imports from a given module on one single "from ... import ..." statement

**Review by Lawouach — DISMISSED** [review]
LGTM

**Comment by Lawouach** [comment]
I wonder, should we have several probe modules per concern rather than a global one?

```
from chaosinstana.events.probes import ...
from chaosinstana.somethingelse.probes import ...
```

vs

```
from chaosinstana.probes import ...
```

File: chaosinstana/api.py (hunk 1 above)
Comment (gtfisher) [reply chaosinstana/api.py]: Is it useful to have the instana response logged to debug?

Comment (russmiles) [reply chaosinstana/api.py]: I can see the need for logging whether the call was successful as info level, maybe response if debug level as you have it?

File: chaosinstana/probes.py (hunk 2 above)
Comment (gtfisher) [reply chaosinstana/probes.py]: datetime is now used to support logging statements

File: chaosinstana/probes.py (hunk 3 above)
Comment (gtfisher) [reply chaosinstana/probes.py]: fixed

File: chaosinstana/probes.py (hunk 8 above)
Comment (gtfisher) [reply chaosinstana/probes.py]: I have changes the logging to debug, I am not sure if this should raise an ActivityFailed, the exception would get raised if when parsing the instana response looking for severity, I have treated it as a if the severity level  was not found - this does not seem correct but not sure how to correct it

File: chaosinstana/probes.py (hunk 7 above)
Comment (gtfisher) [reply chaosinstana/probes.py]: how do I declare a function that returns a boolean, cant seem to find a way the pleases the linter

File: chaosinstana/probes.py (hunk 8 above)
Comment (gtfisher) [reply chaosinstana/probes.py]: amended logging to debug logging

File: chaosinstana/probes.py (hunk 6 above)
Comment (gtfisher) [reply chaosinstana/probes.py]: now debug log millisecond and full date time string

File: chaosinstana/probes.py (hunk 8 above)
Comment (gtfisher) [reply chaosinstana/probes.py]: after discussion with russmiles we decided this was not an ActivityFailed, and just a debug log entry

**Review by russmiles — APPROVED** [review]

**Pull request merged**